napari/layers/utils/style_encoding.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterable, Optional, Sequence, Tuple, Union
+
+import numpy as np
+from pydantic import ValidationError, parse_obj_as, validator
+
+from ...utils import Colormap
+from ...utils.colormaps import ValidColormapArg, ensure_colormap
+from ...utils.colormaps.categorical_colormap import CategoricalColormap
+from ...utils.colormaps.standardize_color import transform_color
+from ...utils.events import EventedModel
+
+
+class StyleEncoding(EventedModel, ABC):
+    array: np.ndarray = []
+
+    @abstractmethod
+    def _apply(
+        self, properties: Dict[str, np.ndarray], indices: Sequence[int]
+    ) -> np.ndarray:
+        pass
+
+    def update_all(self, properties: Dict[str, np.ndarray], n_rows: int):
```

Comment:
This updates the first `n_rows` right?

Reply:
Actually, this is more like a complete refresh that will clobber all the previously generated values.

The best way to think about `n_rows` is that it tells us the total number of rows in the property table described by `properties`. If there are some columns in that table (that have the same size), then we don't need it (because we can do something like `len(next(iter(properties.values()))`). But if there are no columns/properties (i.e. the equivalent DataFrame has shape Nx0), we need it specifically for the constant encodings (and maybe others in the future) to know how many values to generate.

It might be nice if the constant encodings just returned the single 4x1 RGBA vector (which vispy might be able to broadcast), but then we probably need to do some special casing for those encodings. My intuition is that it's easier in the long term to make the constant encodings store more values than they need to, but keep this consistent interface. But I'm open to other ideas.

Reply:
ok - i actually still don't understand this, i think something is slightly off with my mental model/ it was also a long day. Once you've done a round of doc string additions I can read again. I don't want to hold anything up though. `total_rows` might be a better variable name, but i can't quite say. Maybe it will be obvious once it clicks

Reply:
The way I think about it is that I expect to be able to get `n_rows` from `properties`, even when there are no columns (which is the default) and it should correspond to the number of elements (e.g. points) in the owning layer. Given our current dict-of-columns data structure, that's not possible, so we have to pass `n_rows` around with `properties`.

That's required because we always greedily add new values as elements (e.g. points) are added to a layer, even when using a constant or direct encoding and the values aren't actually related to properties.

If we want to avoid that, one alternative is to change the API to have something like `StyleEncoding.get_values(properties, indices)`, which would either generate values on demand or return previously generated values from a cache that is keyed by value (and for constant, just return that constant to be broadcasted as needed).

Reply:
ok yup, i'm getting this now - i like current approach, though generating values on demand could be helpful for scale, but only if all values are constant. seems fine to leave as is